Add support for AWS-LC to openssl and openssl-sys crates

skmcgrail opened the pull request:
This pull request adds support for [AWS Libcrypto](https://github.com/awslabs/aws-lc) (AWS-LC). This is AWS's BoringSSL fork that has been developed with additional performance optimizations and cryptographic algorithm support. The work to integrate AWS-LC into these crates required minimal changes from the existing BoringSSL support.

There are two feature flags that are added depending on the use-case required by the end-consumer. The first feature is `aws-lc` which will depend on our standard AWS-LC sys crate bindings which is based off of tagged release's of AWS-LC's main branch. The other feature is `aws-lc-fips` which will utilize the FIPS version of our sys crate. This currently uses the `fips-2022-11-02` branch of AWS-LC which corresponds to our AWS-LC-FIPS 2.0 module with this associated [certificate](https://csrc.nist.gov/projects/cryptographic-module-validation-program/certificate/4816) and [security policy](https://csrc.nist.gov/CSRC/media/projects/cryptographic-module-validation-program/documents/security-policies/140sp4816.pdf).

skmcgrail commented:
Looks like I might have misjudged one of the BoringSSL function signatures, will fix that in a subsequent commit.

skmcgrail commented:
Hey @sfackler just curious if you've had a chance to consider this pull request? If there are concerns that you may have, or more detail on why we are looking to perform this level of integration please let me know and I would be more then willing to help address those. Thanks!

Skepfyr commented:
How is AWS-LC versioned? How stable is it and how do the bindings line up? Please correct me if this instance is different but taking on a new flavour of OpenSSL makes every single PR harder to write and review because we have to consider this. BoringSSL is odd because it's entirely unstable, is that also true for AWS-LC?

sfackler commented:
I had hoped that the OpenSSL-alike situation had stabilized a decade ago with the post-Heartbleed introduction of LibreSSL and BoringSSL. Unfortunately, in the last year or so it appears that both AWS and Alibaba have decided to create their own new forks (or forks of forks) in AWS-LC and BabaSSL and sought support for them in this library.

Frankly, this is a disappointing trend. The open source ecosystem exists in a robust form today because corporations find it more beneficial to contribute improvements back to upstream open source projects than to maintain their own independent forks. I certainly hope that AWS-LC exists purely as a staging ground for the floating set of patches that have not yet been upstreamed!

Whatever the shape of projects like AWS-LC and BabaSSL actually are, I see three general options for them with respect to rust-openssl:

1. Those projects put in the effort to look enough like their upstream that openssl-sys is happy to link to them as-is.
2. The teams maintaining the forks of OpenSSL maintain forks of this project.
3. If the organizations managing the forks are not interested in dealing with this themselves, they can pay us to formally support their forks.

theoparis commented:
> How is AWS-LC versioned? How stable is it and how do the bindings line up? Please correct me if this instance is different but taking on a new flavour of OpenSSL makes every single PR harder to write and review because we have to consider this. BoringSSL is odd because it's entirely unstable, is that also true for AWS-LC?

Is this still being worked on? AWS-LC is supposed to be more stable than boringssl since it has release tags.

alex commented:
I spoke with @skmcgrail the other day, and I think it makes sense for us to add support for aws-lc at this point. It's gained a fair bit of traction in other parts of the rust ecosystem (notably rustls), and so there's value in being able to use the same crypto library consistently.

Separately, I think if we raised our minimum OpenSSL/LibreSSL version, we'd be able to drop a ton of `cfg()` and reduce our overall maintenance burden.

I'm happy to review this, so let me know when its ready.

alex commented on openssl-sys/src/ocsp.rs:
```diff
@@ -10,18 +10,18 @@ pub const OCSP_REVOKED_STATUS_CESSATIONOFOPERATION: c_int = 5;
-pub const OCSP_TRUSTOTHER: c_ulong = 0x200;
-pub const OCSP_RESPID_KEY: c_ulong = 0x400;
-pub const OCSP_NOTIME: c_ulong = 0x800;
+pub const OCSP_NOCERTS: c_int = 0x1;
```
Hmmm, kind of technically this is backwards incompatible I guess? Why is the change needed?

skmcgrail replied:
Let me investigate further, I forgot about making this change. I think this is a difference in how bindgen is interpreting the preprocessor `#defines` in our sys crate and picking a c_int rather then a c_ulong. This is related to the comment further below as well.

botovq replied:
bindgen is right to choose a `c_int` here since these preprocessor constants are all representable as a `c_int` on almost all platforms (even if c_int happens to be `i16`), so their type in C is `c_int` (e.g. per the table in "The type of the integer constant" in [cppreference][1] - that's 6.4.4.1 in most C standards).

The `c_ulong` will have been chosen for simplicity since their primary use is as a flags argument to `OCSP_basic_verify()` (and a few other OCSP APIs that rust-openssl doesn't have bindings for).

[1]: https://en.cppreference.com/w/c/language/integer_constant

alex commented on openssl/src/ocsp.rs:
```diff
@@ -17,7 +17,7 @@ use openssl_macros::corresponds;
     #[derive(Copy, Clone, Debug, Eq, Hash, Ord, PartialEq, PartialOrd)]
     #[repr(transparent)]
-    pub struct OcspFlag: c_ulong {
+    pub struct OcspFlag: c_int {
```
I believe this is also technically backwards incompatible. What's the motivation?

alex commented on openssl/src/ssl/test/mod.rs:
```diff
@@ -309,17 +309,19 @@ fn pending() {
-        s.ssl().state_string_long(),
-        "SSL negotiation finished successfully"
-    );
+    #[cfg(awslc)]
+    assert_eq!(s.ssl().state_string(), EXPECTED_STATE_STRING_LONG);
+
+    assert_eq!(s.ssl().state_string_long(), EXPECTED_STATE_STRING_LONG);
```
What's going on here?

skmcgrail replied:
AWS-LC delegates `state_string()` down to `state_string_long()` so the output is 1-1 the same. Where as BoringSSL does not do that and just output the fixed string that you were testing for.

alex reviewed (commented):
A few comments, but mostly this looks ok.

In addition: Let's add this to CI so it doesn't regress

skmcgrail commented on openssl-sys/build/main.rs:
```diff
@@ -71,6 +71,47 @@ fn check_ssl_kind() {
+            }
+        };
+
+        let mut version = None;
+        for (name, _) in std::env::vars() {
+            if let Some(name) = name.strip_prefix(env_var_prefix) {
+                if let Some(name) = name.strip_suffix("_INCLUDE") {
+                    version = Some(name.to_owned());
+                    break;
+                }
+            }
+        }
+        let version = version.expect("aws-lc version detected");
```
Will add a comment to the code.

skmcgrail commented on openssl-sys/src/lib.rs:
```diff
@@ -29,6 +29,22 @@ mod boringssl {
+    extern crate aws_lc_sys as aws_lc;
+    pub use aws_lc::*;
+
+    // TODO: AWS-LC doesn't currently expose this in it's public headers
+    extern "C" {
+        pub fn OCSP_ONEREQ_free(r: *mut OCSP_ONEREQ);
+    }
```
This should be fairly easy to expose, I'm not sure what the motivation was originally other then simplifying the surface area of the OCSP APIs. I'll open a pull request to AWS-LC to get these exposed as it would increase our API compatibility with OpenSSL overall.

skmcgrail commented on openssl-sys/build/main.rs:
```diff
@@ -71,6 +71,47 @@ fn check_ssl_kind() {
+    }
 }
 
 fn main() {
```
How important is it that a rust-openssl customer can bring their own pre-compiled cryptographic library like OpenSSL, BoringSSL, and in this case AWS-LC and generate bindings on that? I think in the OpenSSL case this use case makes sense, as in the case of Linux the package manager is handling providing pre-compiled artifacts, updates/patches etc. In the case of BoringSSL, there is no published sys-crate available with bindings already produced, so the pre-compiled route is really the only way to go. (Or build BoringSSL and the create yourself, and then point at it using `.cargo/config.toml`).

I think where AWS-LC differs here is that we have a supported and published bindings crate that we publish for both our FIPS and non-FIPS offerings of the library, and to date don't have AWS-LC distributed in an operating systems package manager. While there is no reason we couldn't add the support here necessary to make it work, I am wondering what the value add would be in the majority use-case? This is ability that we don't actually support in our sys crates either today, and to do it here would also require maintaining another location for our bindgen logic.

What were your thoughts in this space?

alex replied:
For Rust consumers of this library, I kind of suspect it doesn't matter terribly much.

For pyca/cryptography we have the interesting (and frustrating) situation that we'd like users to be able to `pip install` a source package and then build it against whatever OpenSSL they'd like... and we don't have a way to control the cargo features there! So that's really the primary motivation driving me caring about `OPENSSL_DIR` with generality.

(FWIW, it _does_ work for BoringSSL today, via bindgen.)

justsmth commented on .github/workflows/ci.yml:
```diff
@@ -274,18 +276,25 @@ jobs:
               OS_COMPILER=linux-elf
               OS_FLAGS="-m32 -msse2"
               ;;
             "arm-unknown-linux-gnueabihf")
+              CPU=armv4
```
According to the notes [here](https://doc.rust-lang.org/beta/rustc/platform-support.html#tier-2-with-host-tools), this target is `Armv6 Linux, hardfloat (kernel 3.2, glibc 2.17)`. 

So, "armv6"?

skmcgrail replied:
This isn't the Rust target definition in this case, this is the compiler being used to precompile the crypto library, so either aws-lc, openssl, or boringssl. The line immediately below this one indicates its an armv4 compiler being used?

skmcgrail commented:
Rebased and appeared to have broken some things, need to get in a bit more. Will ping this when it's ready to look at.

skmcgrail commented:
All the checks are passing with our latest sys crate release, should be good for eyes again.

alex commented:
Will add to my TODO list once I'm not sick, thanks.

alex commented on openssl/src/ocsp.rs:
```diff
@@ -18,17 +18,17 @@ bitflags! {
+        const NO_CHECKS = ffi::OCSP_NOCHECKS as c_ulong;
+        const TRUST_OTHER = ffi::OCSP_TRUSTOTHER as c_ulong;
+        const RESPID_KEY = ffi::OCSP_RESPID_KEY as c_ulong;
+        const NO_TIME = ffi::OCSP_NOTIME as c_ulong;
```
So the implication here is that `ffi::` types don't match between aws-lc and other OpenSSL-variants?

Is that an issue for anything else? I guess not?

alex replied:
Gotcha.

alex reviewed (commented):
Very very close, final mile!

Before this is mergable, I want to make sure @sfackler has no objections, but modulo this last comment, this LGTM.

alex reviewed (approved):
This LGTM to me. Pinging @sfackler to make sure he's ok too.

alex commented:
FWIW, at this juncture I think it'd be much easier to merge this PR as-is and expand the exposed functionality in follow-up PRs.

skmcgrail commented:
Pushed one more commit to enable a few additional capabilities that we can technically support from the AWS-LC side. Added a few tests as well on the cipher side for some of the AES modes.

alex commented:
Looks like CI is failing.

skmcgrail commented:
Sad, the remaining failures appear to be around the test cases for AES-128-XTS and AES-256-XTS that I added. I assume there might be some API behavioral differences in the older versions of OpenSSL and for LibreSSL that don't align with how the tests are structured. I can try to investigate why that is happening and what can be changed. I added in the tests since this was a gap in coverage at the moment.

alex commented:
I would favor reverting the latest set of additions, landing as is, and
then you can pursue exposing the additional functionality in follow-up PRs
that can all be merged individually.

On Tue, Mar 18, 2025 at 4:45 PM Sean McGrail ***@***.***>
wrote:

> Sad, the remaining failures appear to be around the test cases for
> AES-128-XTS and AES-256-XTS that I added. I assume there might be some API
> behavioral differences in the older versions of OpenSSL and for LibreSSL
> that don't align with how the tests are structured. I can try to
> investigate why that is happening and what can be changed. I added in the
> tests since this was a gap in coverage at the moment.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/sfackler/rust-openssl/pull/1805#issuecomment-2734684036>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAAGBBDBX3L3G2CZPC7D4D2VCAXZAVCNFSM6AAAAABYH6KFYKVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDOMZUGY4DIMBTGY>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
> [image: skmcgrail]*skmcgrail* left a comment (sfackler/rust-openssl#1805)
> <https://github.com/sfackler/rust-openssl/pull/1805#issuecomment-2734684036>
>
> Sad, the remaining failures appear to be around the test cases for
> AES-128-XTS and AES-256-XTS that I added. I assume there might be some API
> behavioral differences in the older versions of OpenSSL and for LibreSSL
> that don't align with how the tests are structured. I can try to
> investigate why that is happening and what can be changed. I added in the
> tests since this was a gap in coverage at the moment.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/sfackler/rust-openssl/pull/1805#issuecomment-2734684036>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAAGBBDBX3L3G2CZPC7D4D2VCAXZAVCNFSM6AAAAABYH6KFYKVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDOMZUGY4DIMBTGY>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>


-- 
All that is necessary for evil to succeed is for good people to do nothing.

botovq commented:
Not sure if my comment is visible anywhere, soI repeat it here. That test breakage in CCM mode on old OpenSSL and LibreSSL is an old trap:

```rust
        let mut buf = vec![];
        ctx.cipher_update(&aad, None).unwrap();
        ctx.cipher_update_vec(&ct, &mut buf).unwrap();
        ctx.cipher_final_vec(&mut buf).unwrap();
```

EVP_DecryptFinal_ex() must not be called when decrypting in CCM mode. This is pointed out in
["Authenticated Decryption using CCM mode" in the old OpenSSL wiki][1]. Skipping that `ctx_cipher_final_vec()` on line 902 for (what will be) LibreSSL 4.1.0 makes tests pass, so that should be easily handled with a `cfg_if!`.


[1]: https://wiki.openssl.org/index.php/EVP_Authenticated_Encryption_and_Decryption#Authenticated_Decryption_using_CCM_mode

skmcgrail commented:
> Not sure if my comment is visible anywhere, so I repeat it here. That test breakage in CCM mode on old OpenSSL and LibreSSL is an old trap:
> 
> ```rust
>         let mut buf = vec![];
>         ctx.cipher_update(&aad, None).unwrap();
>         ctx.cipher_update_vec(&ct, &mut buf).unwrap();
>         ctx.cipher_final_vec(&mut buf).unwrap();
> ```
> 
> EVP_DecryptFinal_ex() must not be called when decrypting in CCM mode. This is pointed out in ["Authenticated Decryption using CCM mode" in the old OpenSSL wiki](https://wiki.openssl.org/index.php/EVP_Authenticated_Encryption_and_Decryption#Authenticated_Decryption_using_CCM_mode). Skipping that `ctx_cipher_final_vec()` on line 902 for (what will be) LibreSSL 4.1.0 makes tests pass, so that should be easily handled with a `cfg_if!` skipping this for the affected versions.

Thanks for tracking this down, very much appreciated. Will have a subsequent PR with that commit here shortly including this suggested fix.
